### PR TITLE
Add sos cmd

### PR DESCRIPTION
**Pull request by pierre-emmanuelJ**
Signed-off-by: Pierre-Emmanuel Jacquier <pierre-emmanuel.jacquier@epitech.eu>

### CODE REVIEW

**Comment by pierre-emmanuelJ** [comment]
RABASED on exo-config

**Comment by greut** [comment]
If you plan to expose the `endpoint` string in the config, as an option, using the go `text/template` feels safe. Probably overkill though.

```golang
package main

import (
	"fmt"
	"bytes"
	"text/template"
)

func main() {
	var b bytes.Buffer

	tpl, _ := template.New("sosEndpoint").Parse("https://sos-{{ . }}.exo.io")
	_ = tpl.Execute(&b, "de-fra-1")

	// or

	tpl, _ := template.New("sosEndpoint").Parse("https://sos-{{ .Zone }}.exo.io")
	_ = tpl.Execute(&b, struct{Zone string}{"de-fra-1"})

	fmt.Println(b.String())
}
```

File: cmd/exo/cmd/sos_list.go [hunk 1]
```diff
@@ -88,26 +97,30 @@ var sosListCmd = &cobra.Command{
 			lastModified := fmt.Sprintf("%v", message.LastModified)
 			key := filepath.ToSlash(message.Key)
 			key = strings.TrimLeft(message.Key[len(prefix):], "/")
-			fmt.Printf("[%s]    %s %s\n", lastModified, size, key)
+			table.AppendArgs(fmt.Sprintf("[%s]    ", lastModified), size, key)
```
Comment (greut) [inline cmd/exo/cmd/sos_list.go]: why the spaces?

File: cmd/exo/table/table.go [hunk 2]
```diff
@@ -47,3 +47,15 @@ func (t *Table) Render() {
 		t.Table.Render()
 	}
 }
+
+// RemoveFrame remove all border and separator
+func (t *Table) RemoveFrame() {
+	t.SetBorder(false)
+	t.SetColumnSeparator("")
+	tablewriter.PadLeft("", "", 0)
+}
```
Comment (greut) [inline cmd/exo/table/table.go]: you endup with that one:

https://golang.org/pkg/text/tabwriter/

Comment (pierre-emmanuelJ) [reply cmd/exo/table/table.go]: thx

File: cmd/exo/cmd/sos_list.go (hunk 1 above)
Comment (pierre-emmanuelJ) [reply cmd/exo/cmd/sos_list.go]: lisibility
```
[2018-07-24 08:10:55.334 +0000 UTC]      0B  cloudstack.ini
[2018-07-23 14:25:24.253 +0000 UTC]      0B  logo
[2018-07-23 16:08:47.031 +0000 UTC]      0B  test
[2018-07-24 08:16:26.758 +0000 UTC]      0B  toto/
```

Comment (pierre-emmanuelJ) [reply cmd/exo/cmd/sos_list.go]: like mc

File: cmd/exo/cmd/sos_acl.go [hunk 3]
```diff
@@ -105,6 +106,14 @@ var sosAddACLCmd = &cobra.Command{
 	},
 }
 
+//merge src header in dst header
+func mergeHeader(src, dst http.Header) http.Header {
+	for k, v := range src {
+		dst[k] = v
+	}
+	return dst
+}
```
Comment (greut) [inline cmd/exo/cmd/sos_acl.go]: This this highly confusing.
- `src`, `dst` vs <https://golang.org/pkg/builtin/#copy>
- you modify `dst` **and** return it

File: cmd/exo/cmd/sos_headers.go [hunk 4]
```diff
@@ -172,14 +172,15 @@ var sosRemoveHeadersCmd = &cobra.Command{
 		}
 
 		objInfo.Metadata["content-type"] = []string{objInfo.ContentType}
+		objInfo.Metadata["x-amz-metadata-directive"] = []string{"REPLACE"}
```
Comment (greut) [inline cmd/exo/cmd/sos_headers.go]: I'd use the appropriate API of `net.http.Header`

https://golang.org/pkg/net/http/#Header

**Comment by greut** [comment]
> cmd/exo/cmd/sos.go:13:31:warning: "managment" is a misspelling of "management" (misspell)
>
> cmd/exo/cmd/sos_acl.go:35:24:warning: "managment" is a misspelling of "management" (misspell)
>
> cmd/exo/cmd/sos_list.go:31:52:warning: "suport" is a misspelling of "support" (misspell)
>
> cmd/exo/cmd/sos_list.go:31:15:warning: error return value not checked (fmt.Fprintf(os.Stderr, "XXX waiting for all zone suport SOS: current zone is %q. use a supported defaultZone", gCurrentAccount.DefaultZone)) (errcheck)

File: cmd/exo/cmd/sos_acl.go (hunk 3 above)
Comment (pierre-emmanuelJ) [reply cmd/exo/cmd/sos_acl.go]: I want to merge two header. parameters order is wrong ?

**Comment by greut** [comment]
I have this

```toml
[[accounts]]
  account = "yoan.blanc@exoscale.ch"
  sosEndpoint = "https://ppsos-{zone}.exo.io"
```

and `exo sos ls` works, much nicer!!

**Comment by greut** [comment]
meh?

```console
% exo sos ls yoanyoanyoan                                                                    
 [2018-07-03 09:59:48 UTC]    0 B a b c/
% exo sos ls "yoanyoanyoan/a b c"
 [2018-07-03 10:00:09 UTC] 89 KiB Screenshot from 2018-06-28 16-10-55.png

% exo sos remove yoanyoanyoan "a b c"                                                        
Object(s):
- a b c
Successfully removed

% exo sos ls "yoanyoanyoan/a b c"    
 [2018-07-03 10:00:09 UTC] 89 KiB Screenshot from 2018-06-28 16-10-55.png
```

**Comment by pierre-emmanuelJ** [comment]
wwwuuuuuttttt

**Comment by pierre-emmanuelJ** [comment]
for this :
image

remove your config file and try this endpoint is hidden

**Comment by greut** [comment]
> remove your config file and try this endpoint is hidden

I meant that it works, which is the goal.

**Comment by greut** [comment]
the good, it works as expected with a folder.

```console
% exo sos remove yoanyoanyoan "a b c"
a b c: is a directory
% exo sos remove yoanyoanyoan "a b c" -r
```

the bad, it doesn't complain that the resource doesn't exist.

```console
% exo sos remove yoanyoanyoan "a b c"

% rm "a b c"
rm: cannot remove 'a b c': No such file or directory
```

is it easy?

**Comment by pierre-emmanuelJ** [comment]
I think it can be possible

**Comment by pierre-emmanuelJ** [comment]
so you can try

File: cmd/exo/cmd/sos_remove.go [hunk 5]
```diff
@@ -82,7 +82,7 @@ var removeCmd = &cobra.Command{
 					nbFile++
 				}
 				if nbFile == 0 {
-					fmt.Fprintf(os.Stderr, "rm: cannot remove '%s': No such file or directory\n", arg)
+					fmt.Fprintf(os.Stderr, "rm: cannot remove '%s': No such file or directory\n", arg) // nolint: errcheck
```
Comment (greut) [inline cmd/exo/cmd/sos_remove.go]: this doesn't return a proper error code

Comment (pierre-emmanuelJ) [reply cmd/exo/cmd/sos_remove.go]: https://github.com/exoscale/egoscale/pull/182/commits/17cbbb05a50b3a58113dcd842ff39114ef873803

File: cmd/exo/cmd/sos_acl.go (hunk 3 above)
Comment (greut) [reply cmd/exo/cmd/sos_acl.go]: - `copy` uses `dst, src`
- coming from the Python world, you _either_ modify and returns `None`, or retuns the modified value.

**Comment by pierre-emmanuelJ** [comment]
Thanks @brutasse 
@greut as your wish 😉

File: cmd/exo/cmd/sos_upload.go [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+	Short:   "Upload an object into a bucket",
+	Aliases: gUploadAlias,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		if len(args) < 2 {
+			return cmd.Usage()
+		}
+
+		var remoteFilePath string
+		if len(args) > 2 {
+			remoteFilePath = args[2]
```
[the remark this answers is not in the document]
Comment (pierre-emmanuelJ) [reply cmd/exo/cmd/sos_upload.go]: yep I did wrong !

File: cmd/exo/cmd/sos_upload.go [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+		}
+
+		if remoteFilePath == "" {
+			remoteFilePath = filePath
+		}
+
+		contentType := http.DetectContentType(buffer)
+
+		// Upload object with FPutObject
+		n, err := minioClient.FPutObjectWithContext(gContext, bucketName, remoteFilePath, filePath, minio.PutObjectOptions{ContentType: contentType})
```
Comment (greut) [inline cmd/exo/cmd/sos_upload.go]: https://github.com/minio/minio-go/blob/2508a98a820e993b0307bbac4db9731e4e6f89a0/api-put-object.go#L37

By bored to death, I meant that it should show a progress bar or something along those lines.

Comment (pierre-emmanuelJ) [reply cmd/exo/cmd/sos_upload.go]: okay thanks

File: cmd/exo/cmd/sos_upload.go [hunk 8]
```diff
@@ -65,19 +69,59 @@ var sosUploadCmd = &cobra.Command{
...
+
+		bar := progress.AddBar((fileInfo.Size()),
+			mpb.PrependDecorators(
+				// display our name with one space on the right
+				decor.Name(objectName, decor.WC{W: len(objectName) + 1, C: decor.DidentRight}),
+				// replace ETA decorator with "done" message, OnComplete event
+				decor.OnComplete(
+					// ETA decorator with ewma age of 60, and width reservation of 4
+					decor.EwmaETA(decor.ET_STYLE_GO, 60, decor.WC{W: 4}), "done",
+				),
```
Comment (greut) [inline cmd/exo/cmd/sos_upload.go]: ETA shows 0s, 1s, 2s on a big file. Where does it get its data from?

File: cmd/exo/cmd/request.go [hunk 9]
```diff
@@ -48,13 +48,15 @@ func asyncTasks(tasks []task) ([]interface{}, []error) {
...
-			),
+
+			//XXX remove for display issue
+			// mpb.AppendDecorators(
+			// 	// replace ETA decorator with "done" message, OnComplete event
+			// 	decor.OnComplete(
+			// 		// ETA decorator with ewma age of 60
+			// 		decor.EwmaETA(decor.ET_STYLE_GO, 60), "done",
+			// 	),
+			// ),
```
Comment (greut) [inline cmd/exo/cmd/request.go]: then remove it.

**Comment by pierre-emmanuelJ** [comment]
https://github.com/vbauerster/mpb/issues/26

File: cmd/exo/cmd/request.go (hunk 9 above)
Comment (pierre-emmanuelJ) [reply cmd/exo/cmd/request.go]: https://github.com/vbauerster/mpb/issues/26

**Pull request merged**